helix-core/src/main/java/org/apache/helix/controller/stages/MessageGenerationPhase.java

```diff
@@ -194,7 +194,8 @@ private void generateMessage(final Resource resource, final BaseControllerDataPr
         }
 
         if (pendingMessage != null && shouldCleanUpPendingMessage(pendingMessage, currentState,
-            currentStateOutput.getEndTime(resourceName, partition, instanceName))) {
+            currentStateOutput.getEndTime(resourceName, partition, instanceName)) && sessionIdMap
+            .containsKey(instanceName)) {
```

Comment:
I would suggest to move this check in side function of shouldCleanUpPendingMessage. This could make the all the check logic inside shouldCleanUpPendingMessage(.